### ISSUE TITLE
PS256 not working on Android

### ISSUE THREAD

**Issue by cgal-capco** [jwtk/jjwt#724]
**Describe the bug**
After following the steps in [Android installation steps](https://github.com/jwtk/jjwt#install-android), and being sure Bounce Castle is in the classpath, when signing a token using PS256 it throws an error saying "Unavailable RSA Signature algorithm 'RSASSA-PSS'.

It looks `SHA256withRSAandMGF1` it's available , but not `RSASSA-PSS`

**Trace**
```
io.jsonwebtoken.security.SignatureException: Unavailable RSA Signature algorithm 'RSASSA-PSS'.
        at io.jsonwebtoken.impl.crypto.SignatureProvider.createSignatureInstance(SignatureProvider.java:68)
        at io.jsonwebtoken.impl.crypto.RsaProvider.createSignatureInstance(RsaProvider.java:69)
        at io.jsonwebtoken.impl.crypto.RsaSigner.doSign(RsaSigner.java:53)
        at io.jsonwebtoken.impl.crypto.RsaSigner.sign(RsaSigner.java:43)
        at io.jsonwebtoken.impl.crypto.DefaultJwtSigner.sign(DefaultJwtSigner.java:59)
        at io.jsonwebtoken.impl.DefaultJwtBuilder.compact(DefaultJwtBuilder.java:353)
```

**Comment by bdemers** [comment]
I don't spend much time with Android, so someone else might have to chime in.
But to try to point you in the right direction, you can set a breakpoint here:

https://github.com/jwtk/jjwt/blob/a4130dd1ec5d4bef4b1835927a2a1d966a519109/api/src/main/java/io/jsonwebtoken/lang/RuntimeEnvironment.java#L34

And check to make sure the BouncyCastle Security Provider was installed.

You can also set a break point in the SignatureProvider from your stack trace and use your debugger to evaluate [`Security.getProviders()`](https://docs.oracle.com/javase/7/docs/api/java/security/Security.html#getProviders()).

Also, this answer from SO implies you may need to watch out for the Android version of BC as well:
https://stackoverflow.com/a/66323575/6415199

If I'm way off here please let us know!

If you are still stuck, if you could create a small mini project that we can reproduce the problem that would be great! (let us know the version of Java you are using too)

**Comment by lhazlewood** [comment]
Just a note for reference.  JJWT currently will auto-load and install BC as a SecurityProvider if found in the classpath when JJWT loads, but this 'pollutes' the system environment with something that the app developer really should do themselves if they need it for their purposes.

The `jwe` branch has changed logic to:
  1. *only* load BC if it is actually needed based on a runtime condition, e.g. https://github.com/jwtk/jjwt/blob/3ba22feb0ee1655ffe0f08aac88c996f15ba11c1/impl/src/main/java/io/jsonwebtoken/impl/security/Providers.java#L64
  AND
  2. keep the loaded BC provider local to JJWT's internal needs, _without_ calling `Security.addProvider` so as to not impact the application's system environment.

**Comment by cgal-capco** [comment]
Hi! Thanks for the suggestions.

I've managed to fix the problem with the `RSASSA-PSS` support. As reference, I had to remove the BC provider and add it again.

```
    Security.removeProvider(BouncyCastleProvider.PROVIDER_NAME);
    Security.addProvider(BouncyCastleProvider());
```

Now I've got another problem. 

I've expend some time debugging this and it looks the problem it's that the Private key it's a `AndroidKeyStoreRSAPrivateKey`, and BC expects a `RSAPrivateKey`.

I'm generating the RSA keys using the keystore as per the android documentation (I need to set the alias and some other config).
